packages/workspace/src/node/default-workspace-server.ts

```diff
@@ -24,36 +24,60 @@ import { CliContribution } from '@theia/core/lib/node/cli';
 import { WorkspaceServer, CommonWorkspaceUtils } from '../common';
 import { EnvVariablesServer } from '@theia/core/lib/common/env-variables';
+import URI from '@theia/core/lib/common/uri';
 
 @injectable()
 export class WorkspaceCliContribution implements CliContribution {
 
+    @inject(EnvVariablesServer) protected readonly envVariablesServer: EnvVariablesServer;
+    @inject(CommonWorkspaceUtils) protected readonly workspaceUtils: CommonWorkspaceUtils;
+
     workspaceRoot = new Deferred<string | undefined>();
 
     configure(conf: yargs.Argv): void {
-        conf.usage('$0 [workspace-directory] [options]');
+        conf.usage('$0 [workspace-directories] [options]');
         conf.option('root-dir', {
             description: 'DEPRECATED: Sets the workspace directory.',
         });
     }
 
-    setArguments(args: yargs.Arguments): void {
-        let wsPath: string | undefined = typeof args._[2] === 'undefined' ? undefined : String(args._[2]);
-        if (!wsPath) {
-            wsPath = args['root-dir'] as string;
-            if (!wsPath) {
-                this.workspaceRoot.resolve(undefined);
-                return;
-            }
+    async setArguments(args: yargs.Arguments): Promise<void> {
+        const workspaceArguments = args._.slice(2).map(probablyAlreadyString => String(probablyAlreadyString));
+        if (workspaceArguments.length === 0 && args['root-dir']) {
+            workspaceArguments.push(String(args['root-dir']));
         }
-        if (!path.isAbsolute(wsPath)) {
-            const cwd = process.cwd();
-            wsPath = path.join(cwd, wsPath);
+        if (workspaceArguments.length === 0) {
+            this.workspaceRoot.resolve(undefined);
+        } else if (workspaceArguments.length === 1) {
+            this.workspaceRoot.resolve(this.normalizeWorkspaceArg(workspaceArguments[0]));
+        } else {
+            this.workspaceRoot.resolve(this.buildWorkspaceForMultipleArguments(workspaceArguments));
         }
-        if (wsPath && wsPath.endsWith('/')) {
-            wsPath = wsPath.slice(0, -1);
+    }
+
+    protected normalizeWorkspaceArg(raw: string): string {
+        return path.resolve(raw).replace(/\/$/, '');
+    }
+
+    protected async buildWorkspaceForMultipleArguments(workspaceArguments: string[]): Promise<string | undefined> {
+        try {
+            const dirs = await Promise.all(workspaceArguments.map(async maybeDir => (await fs.stat(maybeDir).catch(() => undefined))?.isDirectory()));
+            const folders = workspaceArguments.filter((_, index) => dirs[index]).map(dir => ({ path: this.normalizeWorkspaceArg(dir) }));
+            if (folders.length < 2) {
+                return folders[0]?.path;
+            }
+            const untitledWorkspaceUri = await this.workspaceUtils.getUntitledWorkspaceUri(
+                new URI(await this.envVariablesServer.getConfigDirUri()),
+                uri => !fs.existsSync(uri.path.toString()),
```

Comment:
`path.toString()` will not resolve to actual file paths on Windows, making this feature non-functional. Also, we can make this actually async.
```suggestion
                async uri => !await fs.pathExists(uri.path.fsPath()),
```